### PR TITLE
Implement type generics for functions

### PR DESCRIPTION
**Pull request by marcusklaas**
Pretty printing of function types isn't as nice any more since the current implementation does not store its argument types directly. We could store some more information to print more information on the argument and return types, at a cost of course.

### CODE REVIEW

File: crates/ra_hir/src/ty.rs [hunk 1]
```diff
@@ -1311,20 +1337,34 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
...
+                // TODO: we should use turbofish hints like this:
+                // f::<u32>(x)
                 let callee_ty = self.infer_expr(*callee, &Expectation::none());
                 let (param_tys, ret_ty) = match &callee_ty {
-                    Ty::FnPtr(sig) => (&sig.input[..], sig.output.clone()),
+                    Ty::FnPtr(sig) => (sig.input.clone(), sig.output.clone()),
+                    Ty::FnDef { def, substs, .. } => {
+                        let fn_sig = def.signature(self.db);
+                        let ret_ty = self
+                            .make_ty(fn_sig.ret_type(), fn_sig.generics())
```
[the remark this answers is not in the document]
Comment (flodiebold) [reply crates/ra_hir/src/ty.rs]: (by reusing `make_ty`, this actually currently resolves the function parameter in the module of the function we're inferring right now...)

**Comment by marcusklaas** [comment]
Haha, woops. That's not right..

You must be the most patient person on GitHub Florian. Thank you once again for the review. I will make it right.

File: crates/ra_hir/src/code_model_api.rs [hunk 2]
```diff
@@ -388,7 +388,8 @@ pub use crate::code_model_impl::function::ScopeEntryWithSyntax;
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct FnSignature {
     pub(crate) name: Name,
-    pub(crate) params: Vec<TypeRef>,
+    pub(crate) type_params: Arc<GenericParams>,
```
[the remark this answers is not in the document]
Comment (matklad) [reply crates/ra_hir/src/code_model_api.rs]: It's even not clear if we need signature at all... it stores unresolved types, so we basically can just fold it into methods of `Function` itself. That is, `FnSignature` is just a convenient grouping, it's not essential.

For generics, I'd also imagine them being a part of `Function` itself. We probably could use a `GenericsOwner` trait and implement it for hir::Struct, Enum, Function, etc.

**Comment by marcusklaas** [comment]
updated

File: crates/ra_hir/src/ty.rs [hunk 3]
```diff
@@ -586,6 +614,25 @@ impl fmt::Display for Ty {
                     .to_fmt(f)?;
                 write!(f, " -> {}", sig.output)
             }
+            Ty::FnDef {
+                name, substs, sig, ..
+            } => {
+                // don't have access to the param types here :-(
```
Comment (flodiebold) [inline crates/ra_hir/src/ty.rs]: that's not true anymore, right

File: crates/ra_hir/src/ty.rs [hunk 4]
```diff
@@ -1103,6 +1157,32 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
...
         let ty = self.insert_type_vars(ty);
+
+        // try to get generic parameters from the path and add them to the
+        // function type substitutions
+        if let Ty::FnDef { ref def, .. } = ty {
+            let last_seg_bindings = path
+                .segments
+                .last()
+                .and_then(|segment| segment.args_and_bindings.as_ref());
+            if let Some(generic_args) = last_seg_bindings {
```
Comment (flodiebold) [inline crates/ra_hir/src/ty.rs]: I think all this should already be done by `substs_from_path`, and it should be enough to call that directly?

File: crates/ra_hir/src/ty.rs [hunk 5]
```diff
@@ -1330,20 +1408,29 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
                 Ty::Unknown
             }
             Expr::Call { callee, args } => {
+                // TODO: we should use turbofish hints like this:
```
Comment (flodiebold) [inline crates/ra_hir/src/ty.rs]: the TODO comment is outdated now, right?

File: crates/ra_hir/src/ty/tests.rs [hunk 6]
```diff
@@ -594,6 +594,31 @@ fn test() {
     );
 }
 
+#[test]
+fn infer_type_param() {
+    check_inference(
+        "generic_fn",
```
Comment (flodiebold) [inline crates/ra_hir/src/ty/tests.rs]:
```suggestion
        "infer_type_param",
```

File: crates/ra_hir/src/ty.rs (hunk 4 above)
Comment (marcusklaas) [reply crates/ra_hir/src/ty.rs]: You are absolutely right - that saves us a lot of work!

**Pull request merged**